File: apps/windows/TeslaSync.App/feature-views/SentryModeChart.Model.cs
using System.Globalization;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive render branch of the <c>SentryModeChart</c> surface — the native union of the
/// states the web component renders
/// (web/src/features/admin/components/security-access/SentryModeChart.tsx). The web source is a pure
/// presentational component (it takes a single <c>sentryBuckets: SentryDayBucket[]</c> prop and performs no
/// fetching), so the branches are a direct function of the input <see cref="SentryModeChartModel"/> — there
/// is no fetch-driven error / stale / offline branch to reproduce here. The parent Security-and-Access
/// experience owns the query lifecycle (loading / error / stale / offline are handled once for the whole
/// page before any chart is shown), exactly as the web page only renders this chart once the security
/// history has resolved. Every branch maps onto a visible surface; none is ever hidden.
/// </summary>
public enum SentryModeChartState
{
    /// <summary>Initial fetch in flight (the parent is still loading the history) — title + skeleton chrome.</summary>
    Loading,

    /// <summary>Resolved with no day buckets to chart — a friendly empty state, never a blank chart.</summary>
    Empty,

    /// <summary>At least one day bucket is present (web <c>sentryBuckets.length &gt; 0</c>) — the stacked bars.</summary>
    Ready,
}

/// <summary>
/// One day's sentry-mode armed/disarmed tally — the native mirror of the web <c>SentryDayBucket</c> shape in
/// <c>web/src/features/admin/components/security-access/helpers.ts</c>
/// (<c>{ date: string; sentryOn: number; sentryOff: number }</c>). <see cref="Date"/> is the <c>YYYY-MM-DD</c>
/// day key the web's <c>buildSentryBuckets</c> emits; <see cref="SentryOn"/> / <see cref="SentryOff"/> are the
/// number of security snapshots that day with sentry armed / disarmed. Pure data — no WinUI types.
/// </summary>
public sealed record SentryDayBucket(string Date, long SentryOn, long SentryOff);

/// <summary>
/// The render-time data model the <c>SentryModeChart</c> view binds to — the native analogue of the web
/// <c>SentryModeChartProps</c> (<c>sentryBuckets: SentryDayBucket[]</c>), plus the fetch flag the parent
/// supplies so the surface can render its own loading branch (the web parent gates the whole chart behind
/// its query state). The component is presentational: this model carries only the day-bucket series and the
/// loading flag. User-facing labels are resolved from the i18n facade by the projection, not passed in. Pure
/// data — no WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
public sealed record SentryModeChartModel(bool Loading, IReadOnlyList<SentryDayBucket> Buckets)
{
    /// <summary>The initial model: the first fetch is in flight and no buckets have arrived yet.</summary>
    public static SentryModeChartModel Pending { get; } =
        new(true, Array.Empty<SentryDayBucket>());

    /// <summary>A resolved model with no buckets — the empty state.</summary>
    public static SentryModeChartModel Empty { get; } =
        new(false, Array.Empty<SentryDayBucket>());
}

/// <summary>
/// One projected, render-ready stacked column — the native analogue of a single day on the recharts
/// stacked <c>&lt;BarChart&gt;</c>. <see cref="AxisLabel"/> is the short date shown beneath the column (the
/// web <c>XAxis tickFormatter={formatDateShort}</c>); <see cref="SentryOn"/> / <see cref="SentryOff"/> are the
/// raw tallies and <see cref="SentryOnText"/> / <see cref="SentryOffText"/> their grouped display form (the
/// web <c>fmtNumber</c> / <c>Intl.NumberFormat</c>); <see cref="OnRatio"/> / <see cref="OffRatio"/> are each
/// segment's height as a fraction (0..1) of the tallest day's total (so the two segments stack to the column's
/// share of the busiest day, mirroring <c>stackId="sentry"</c>); and <see cref="AutomationName"/> is the spoken
/// "{date}, {on} Sentry On, {off} Sentry Off" the web tooltip conveys. Pure data.
/// </summary>
public sealed record SentryModeChartColumn(
    string Date,
    string AxisLabel,
    long SentryOn,
    long SentryOff,
    string SentryOnText,
    string SentryOffText,
    double OnRatio,
    double OffRatio,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the chart for one input model — the native analogue of what the
/// web <c>SentryModeChart</c> renders inside its <c>GlassPanel</c>. Holds the resolved title + series legend
/// labels (the three <c>t(...)</c> calls from the web source), the empty + loading copy, the active
/// <see cref="State"/>, the projected stacked <see cref="Columns"/>, and the surface
/// <see cref="AutomationName"/>. Pure data so every branch is asserted headlessly.
/// </summary>
public sealed record SentryModeChartDisplay(
    SentryModeChartState State,
    string Title,
    string SentryOnLabel,
    string SentryOffLabel,
    string EmptyMessage,
    string LoadingLabel,
    IReadOnlyList<SentryModeChartColumn> Columns,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="SentryModeChartModel"/> to its <see cref="SentryModeChartDisplay"/> — the
/// native port of web/src/features/admin/components/security-access/SentryModeChart.tsx. The web source is a
/// titled, stacked bar chart of the per-day sentry-armed / sentry-disarmed snapshot counts; this projection
/// reproduces its three <c>t(...)</c> labels (<c>admin.security.sentryChart</c> /
/// <c>admin.security.chart.sentryOn</c> / <c>admin.security.chart.sentryOff</c>), scales each day's two
/// segments to the busiest day's total so they stack faithfully (the web <c>stackId="sentry"</c>), formats the
/// short date axis labels (the web <c>formatDateShort</c>) and the tooltip counts through
/// <see cref="NumberFormatting"/> (the web <c>fmtNumber</c>). The web gate is purely
/// <c>sentryBuckets.length &gt; 0</c>: any day bucket promotes the surface to the charted Ready state, and only
/// a resolved, empty series collapses to the friendly empty state per the surface state contract. Every label
/// resolves through the i18n facade using the same keys the web source feeds into <c>t(...)</c>. No WinUI
/// types — unit-tested without a UI host.
/// </summary>
public static class SentryModeChartProjection
{
    /// <summary>Em-dash shown for a blank / unparseable day key (mirrors the web <c>formatDateShort</c> fallback).</summary>
    private const string EmDash = "\u2014";

    private static readonly CultureInfo EnUs = CultureInfo.GetCultureInfo("en-US");

    /// <summary>
    /// Token brush key for the "Sentry On" stacked segment — the W1 chart token whose colour is exactly the
    /// web's <c>#3b82f6</c> blue, kept as a theme-aware key rather than a hard-coded hex.
    /// </summary>
    public const string SentryOnBrushKey = "TsChartSpeedBrush";

    /// <summary>
    /// Token brush key for the "Sentry Off" stacked segment — the neutral status token (the theme-aware
    /// analogue of the web's <c>#6b7280</c> grey), resolved once from <see cref="StatusResources"/>.
    /// </summary>
    public static string SentryOffBrushKey { get; } = StatusResources.AccentBrushKey(StatusKind.Neutral);

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web <c>sentryBuckets</c> prop + the parent's fetch flag).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    public static SentryModeChartDisplay Project(SentryModeChartModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string title = localizer.GetString("admin.security.sentryChart", "Sentry Mode Activity");
        string sentryOnLabel = localizer.GetString("admin.security.chart.sentryOn", "Sentry On");
        string sentryOffLabel = localizer.GetString("admin.security.chart.sentryOff", "Sentry Off");
        string emptyMessage = localizer.GetString("common.noData", "No data available");
        string loadingLabel = localizer.GetString("common.loading", "Loading");

        IReadOnlyList<SentryModeChartColumn> columns = BuildColumns(model.Buckets, sentryOnLabel, sentryOffLabel);
        SentryModeChartState state = SelectState(model.Loading, columns);

        return new SentryModeChartDisplay(
            State: state,
            Title: title,
            SentryOnLabel: sentryOnLabel,
            SentryOffLabel: sentryOffLabel,
            EmptyMessage: emptyMessage,
            LoadingLabel: loadingLabel,
            Columns: columns,
            AutomationName: BuildAutomationName(state, title, emptyMessage, loadingLabel, columns.Count));
    }

    /// <summary>Branch precedence from the web source's data lifecycle: loading → empty → ready.</summary>
    private static SentryModeChartState SelectState(bool loading, IReadOnlyList<SentryModeChartColumn> columns)
    {
        if (loading)
        {
            return SentryModeChartState.Loading;
        }

        // The web renders the chart whenever `sentryBuckets.length > 0` and the friendly empty state otherwise,
        // so emptiness is purely a function of whether any day bucket exists.
        return columns.Count > 0 ? SentryModeChartState.Ready : SentryModeChartState.Empty;
    }

    private static IReadOnlyList<SentryModeChartColumn> BuildColumns(
        IReadOnlyList<SentryDayBucket> buckets,
        string sentryOnLabel,
        string sentryOffLabel)
    {
        if (buckets.Count == 0)
        {
            return Array.Empty<SentryModeChartColumn>();
        }

        long maxTotal = 0;
        foreach (SentryDayBucket bucket in buckets)
        {
            long total = Clamp(bucket.SentryOn) + Clamp(bucket.SentryOff);
            if (total > maxTotal)
            {
                maxTotal = total;
            }
        }

        var columns = new List<SentryModeChartColumn>(buckets.Count);
        foreach (SentryDayBucket bucket in buckets)
        {
            long on = Clamp(bucket.SentryOn);
            long off = Clamp(bucket.SentryOff);
            string onText = NumberFormatting.Format(on, null, 0);
            string offText = NumberFormatting.Format(off, null, 0);
            double onRatio = maxTotal > 0 ? Math.Clamp(on / (double)maxTotal, 0.0, 1.0) : 0.0;
            double offRatio = maxTotal > 0 ? Math.Clamp(off / (double)maxTotal, 0.0, 1.0) : 0.0;
            string axisLabel = FormatAxisLabel(bucket.Date);

            columns.Add(new SentryModeChartColumn(
                Date: bucket.Date ?? string.Empty,
                AxisLabel: axisLabel,
                SentryOn: on,
                SentryOff: off,
                SentryOnText: onText,
                SentryOffText: offText,
                OnRatio: onRatio,
                OffRatio: offRatio,
                AutomationName: $"{axisLabel}, {onText} {sentryOnLabel}, {offText} {sentryOffLabel}"));
        }

        return columns;
    }

    /// <summary>
    /// Short month/day axis label for a <c>YYYY-MM-DD</c> day key — the native analogue of the web
    /// <c>formatDateShort</c> ("Apr 4"). The day key is formatted directly (no timezone round-trip) so the
    /// label is stable across time zones, following the established en-US <c>DateTimeFormatting</c> contract;
    /// a blank or unparseable key renders the em-dash fallback, exactly as the web formatter does.
    /// </summary>
    private static string FormatAxisLabel(string? date)
    {
        if (string.IsNullOrWhiteSpace(date))
        {
            return EmDash;
        }

        if (DateOnly.TryParseExact(date, "yyyy-MM-dd", CultureInfo.InvariantCulture, DateTimeStyles.None, out DateOnly day))
        {
            return day.ToString("MMM d", EnUs);
        }

        if (DateTimeOffset.TryParse(date, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out DateTimeOffset stamp))
        {
            return stamp.UtcDateTime.ToString("MMM d", EnUs);
        }

        return EmDash;
    }

    private static long Clamp(long value) => value < 0 ? 0 : value;

    private static string BuildAutomationName(
        SentryModeChartState state,
        string title,
        string emptyMessage,
        string loadingLabel,
        int dayCount) => state switch
        {
            SentryModeChartState.Loading => $"{title}. {loadingLabel}",
            SentryModeChartState.Empty => $"{title}. {emptyMessage}",
            _ => $"{title}. {dayCount}",
        };
}

/// <summary>
/// PII-safe diagnostics for the <c>SentryModeChart</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never a day key or a sentry count — so a
/// diagnostics line can never leak a user's security activity. Thread-safe.
/// </summary>
public sealed class SentryModeChartDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public SentryModeChartDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=SentryModeChart</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={SentryModeChartRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>SentryModeChart</c> feature surface — the native mirror of the web
/// component at <c>web/src/features/admin/components/security-access/SentryModeChart.tsx</c>.
/// </summary>
public static class SentryModeChartRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "SentryModeChart";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/SentryModeChart/SentryModeChartModel.kt
// Pure, framework-free model + projection for the Sentry Mode Activity chart feature view — the native
// analogue of everything the web component derives before returning JSX
// (web/src/features/admin/components/security-access/SentryModeChart.tsx). No Compose, no Android, no HTTP:
// every declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// The web component is purely presentational — its parent (the security-access page) builds the
// `SentryDayBucket[]` via `buildSentryBuckets(history)` and passes it down. This file owns the parts the web
// render derives from that prop: the two stacked bar series (`sentryOn` / `sentryOff`), the X-axis day
// labels (web `<XAxis dataKey="date" tickFormatter={formatDateShort} />`), and the integer count formatting
// (web `<YAxis allowDecimals={false} />`). The bucket order is preserved exactly as received (the web helper
// already sorts ascending by date and the chart maps in array order), so the native categorical bar chart
// reads left-to-right in the same order.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/SentryModeChart — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.sentrymodechart

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.graphics.vector.path
import androidx.compose.ui.unit.dp
import io.teslasync.shared.core.diagnostics.Logger
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.Locale

/** Em dash shown when a bucket date is missing or unparseable — the web `formatDateShort` `'—'` fallback. */
internal const val EM_DASH: String = "\u2014"

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object SentryModeChartRegistration {
    /** Stable surface id. */
    const val ID: String = "sentry-mode-chart"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "SentryModeChart"
}

/**
 * One day's Sentry Mode activity tally — the native mirror of the web `SentryDayBucket`
 * (`{ date: string; sentryOn: number; sentryOff: number }`). [date] is the `YYYY-MM-DD` day key,
 * [sentryOn] the count of armed observations that day and [sentryOff] the count of disarmed ones
 * (both non-negative integers, so `Long`).
 */
data class SentryDayBucket(
    val date: String,
    val sentryOn: Long,
    val sentryOff: Long,
)

/**
 * The already-localized chart microcopy the composable reads from the i18n catalog (P1/S10) — the three
 * `admin.security.*` keys the web component resolves via `t(...)`: the panel [title] and the two series
 * legend labels ([sentryOnLabel] / [sentryOffLabel]). The lifecycle-chrome strings (empty / error / retry /
 * offline / freshness) are resolved inline at the Compose boundary, not here, so this holder stays a thin
 * content carrier.
 */
data class SentryModeChartStrings(
    val title: String,
    val sentryOnLabel: String,
    val sentryOffLabel: String,
)

/**
 * The fully projected, render-ready chart inputs — the native analogue of the props the web `<BarChart>`
 * reads from `sentryBuckets`. Pure data (no Compose types) so the projection is unit-tested without a UI
 * host: the composable wraps [sentryOnValues] / [sentryOffValues] into two `ChartSeries`, feeds [xLabels] to
 * the bar chart's bottom axis, and shows the empty state when [isEmpty].
 */
data class SentryModeChartProjectionResult(
    val xLabels: List<String>,
    val sentryOnValues: List<Double>,
    val sentryOffValues: List<Double>,
    val isEmpty: Boolean,
)

/**
 * The pure projection the composable renders — the native mirror of the web component's chart data mapping.
 * Stateless and side-effect-free so it is fully covered by the off-device unit gate.
 */
object SentryModeChartProjection {
    /**
     * Projects the loaded [buckets] into render-ready chart inputs, preserving the received order. Each
     * bucket contributes one X-axis label ([formatDate] of its `date`), one `sentryOn` bar value and one
     * `sentryOff` bar value. Injecting the date formatter keeps this function locale/zone-deterministic for
     * tests (the composable supplies the real localized formatter).
     */
    fun project(
        buckets: List<SentryDayBucket>,
        formatDate: (date: String) -> String,
    ): SentryModeChartProjectionResult =
        SentryModeChartProjectionResult(
            xLabels = buckets.map { formatDate(it.date) },
            // `+ 0.0` widens each Long count to the chart series' Double value type.
            sentryOnValues = buckets.map { it.sentryOn + 0.0 },
            sentryOffValues = buckets.map { it.sentryOff + 0.0 },
            isEmpty = buckets.isEmpty(),
        )

    /** Locale-grouped integer formatting for the Y axis — the web `<YAxis allowDecimals={false} />`. */
    fun formatCount(
        count: Long,
        locale: Locale = Locale.getDefault(),
    ): String = String.format(locale, "%,d", count)
}

/**
 * Tolerant date → localized "short month + day" formatter — the native analogue of the web `formatDateShort`
 * (`toLocaleDateString` with `{ month: 'short', day: 'numeric' }`, e.g. `Apr 4`). Pure (java.time only) so it
 * is unit-tested deterministically with a fixed zone/locale. A blank or unparseable input yields [EM_DASH],
 * exactly like the web helper's invalid-date guard. The bucket key is normally a date-only `YYYY-MM-DD`, but
 * the decode chain also tolerates a full ISO date-time (web `new Date(iso)` accepts both).
 */
object SentryDateFormatting {
    private const val MONTH_DAY_PATTERN = "MMM d"

    fun format(
        date: String,
        zone: ZoneId,
        locale: Locale,
    ): String {
        val localDate = parseDate(date, zone) ?: return EM_DASH
        return DateTimeFormatter.ofPattern(MONTH_DAY_PATTERN, locale).format(localDate)
    }

    // Tolerant decode chain: a date-only `YYYY-MM-DD`, then an offset date-time, then a zoneless local
    // date-time, then an RFC-3339 instant resolved in [zone]. The first that parses wins; none parsing
    // yields the em-dash guard above.
    private val parsers: List<(String, ZoneId) -> LocalDate?> =
        listOf(
            { raw, _ -> tryParse { LocalDate.parse(raw) } },
            { raw, _ -> tryParse { OffsetDateTime.parse(raw).toLocalDate() } },
            { raw, _ -> tryParse { LocalDateTime.parse(raw).toLocalDate() } },
            { raw, zone -> tryParse { Instant.parse(raw).atZone(zone).toLocalDate() } },
        )

    private fun parseDate(
        raw: String,
        zone: ZoneId,
    ): LocalDate? = if (raw.isBlank()) null else parsers.firstNotNullOfOrNull { it(raw, zone) }

    private fun tryParse(block: () -> LocalDate): LocalDate? =
        try {
            block()
        } catch (_: DateTimeParseException) {
            null
        }
}

/**
 * Empty-state glyph mirroring the web `<Activity />` (lucide) icon, drawn as a 24×24 stroked
 * [ImageVector] in the shared monochrome style so it recolors via the `Icon` tint. Kept local to this
 * surface (the mandated allowed-files path) rather than expanding the shared icon set from a feature prompt.
 * The path is the lucide `activity` heartbeat polyline (`M22 12h-4l-3 9L9 3l-3 9H2`).
 */
val SentryActivityGlyph: ImageVector =
    ImageVector
        .Builder(
            name = "SentryActivity",
            defaultWidth = 24.dp,
            defaultHeight = 24.dp,
            viewportWidth = 24f,
            viewportHeight = 24f,
        ).apply {
            path(
                stroke = SolidColor(Color.Black),
                strokeLineWidth = 2f,
                strokeLineCap = StrokeCap.Round,
                strokeLineJoin = StrokeJoin.Round,
            ) {
                moveTo(22f, 12f)
                horizontalLineToRelative(-4f)
                lineToRelative(-3f, 9f)
                lineTo(9f, 3f)
                lineToRelative(-3f, 9f)
                horizontalLineTo(2f)
            }
        }.build()

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [SentryModeChartRegistration.SLUG]
 * (P1/S11). Kept free of Compose so it is unit-tested with a recording [Logger]; the composable calls it
 * from its first-composition effect.
 */
fun recordSentryModeChartOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to SentryModeChartRegistration.SLUG))
}
